src/main/scala/gitbucket/core/controller/RepositoryViewerController.scala

```diff
@@ -526,6 +526,118 @@ trait RepositoryViewerControllerBase extends ControllerBase {
     }
   })
 
+  /**
+    * Upload file to a branch.
+    */
+  post("/:owner/:repository/files/upload")(collaboratorsOnly { repository =>
+    defining(repository.owner, repository.name){ case (owner, name) =>
+      (for {
+        data <- extractFromJsonBody[UploadFiles] if data.isValid
+      } yield {
+        Directory.getAttachedDir(owner, name) match {
+          case dir if (dir.exists && dir.isDirectory) =>
+            val _commitFiles = data.fileIds.map { case (fileName, id) =>
+              dir.listFiles.find(_.getName.startsWith(id + ".")).map { file =>
+                val s = scala.io.Source.fromFile(file) // Codec ????
+              val byteArray = s.map(_.toByte).toArray
+
+                CommitFile(id, fileName, byteArray)
+              }
+            }.toList
+
+            val finalCommitFiles = _commitFiles.flatten
+            if(finalCommitFiles.size == data.fileIds.size) {
+              commitFiles(
+                repository,
+                files = finalCommitFiles,
+                branch = data.branch,
+                path = data.path,
+                message = data.message)
+            }
+            else {
+              org.scalatra.NotAcceptable(
+                s"""{"message":
+                    |"$repository doesn't contain all the files you specified in the body"}""".stripMargin)
+            }
+
+          case _ => org.scalatra.NotFound(s"""{"message": "$repository doesn't contain any attached files"}""")
+        }
+
+      }) getOrElse
+        org.scalatra.NotAcceptable("""{"message": "FileIds can't be an empty list"}""")
+
+    }
+  })
+
+  /*
+Roy Li modification
+ */
+
+  case class CommitFile(fileId: String, name: String, fileBytes: Array[Byte])
+
+  private def commitFiles(repository: RepositoryService.RepositoryInfo,
```

Comment:
Existing `RepositoryViewerController.commitFile()` might be possible to be replaced with this method.